packages/contracts/contracts/discovery/ProjectVoting.sol

```diff
@@ -0,0 +1,295 @@
+// SPDX-License-Identifier: UNLICENSED
+pragma solidity =0.8.12;
+
+import {ICommon} from "./interfaces/ICommon.sol";
+import {Math} from "../libraries/Math.sol";
+
+import "hardhat/console.sol";
+
+abstract contract ProjectVoting is ICommon {
+    enum ProjectStatus {
+        InProgress,
+        Won,
+        Lost
+    }
+
+    struct SortableVote {
+        uint256 originalIndex;
+        uint32 count;
+    }
+
+    /// number of votes for each project
+    uint32[] public votes;
+
+    /// Ordered list of slot winners
+    /// @dev only valid up to the last time there was a vote. Use getWinners() instead.
+    address[] winners;
+
+    /// index in the votes array to project address
+    mapping(uint256 => address) public votesIndexToProject;
+
+    /// project address to index  in the votes array
+    mapping(address => uint256) public projectToVotesIndex;
+
+    /// user => votes
+    mapping(address => uint256) public userVoteCount;
+
+    /// project address => votes
+    mapping(address => uint256) public projectVoteCount;
+
+    /// project address => weightedVotes
+    mapping(address => uint256) public weightedProjectVoteCount;
+
+    /// project => (user => voted)
+    mapping(address => mapping(address => bool)) public userVotesPerProject;
```

Comment:
can we rename this to `userHasVotedForProject`? the current name makes it look like it's a counter, not a bool